### PR TITLE
Vllm integration -- Mirage side modification

### PR DESCRIPTION
**Pull request by NorthmanPKU**
**Description of changes:**

1. Reorganize the python/mpk dir
2. Add `GraphBuilder` to build graph based on model name, and, in the future on computation graph
3. Add wrapper `MPK` class above `PersistentKernel` and `GraphBuilder` to provide a better interface to use mpk
4. Add a demo `demo/qwen3/demo_mpk_wrapper.py`.
5. Get rid of self-maintained transformer files.
6. Use stream, event and streamwait instead of devicesync for worker and scheduler

Now we support to run as vllm backend with model name specified.

**Related Issues:**

Linked Issues:
- Issue #522 #342 https://github.com/vllm-project/vllm/issues/22201

Issues closed by this PR:
- Closes #

### CODE REVIEW

File: python/mirage/mpk/mpk.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+            mpi_rank=self.rank,
+            num_workers=self.num_workers,
+            num_local_schedulers=self.num_schedulers,
+            num_remote_schedulers=0,
+            max_seq_length=self.max_seq_length,
+            max_num_batched_requests=args.max_num_batched_requests,
+            max_num_batched_tokens=self.max_num_batched_tokens,
+            max_num_pages=args.max_num_pages,
+            page_size=args.page_size,
+            eos_token_id=151645,
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply python/mirage/mpk/mpk.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/mirage-project/mirage/pull/537#r2513106968)

File: python/mirage/mpk/mpk.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
```
Comment (jiazhihao) [inline python/mirage/mpk/mpk.py]: Let's not hardcode eos_token_id

File: python/mirage/mpk/models/qwen3/modeling_qwen3.py
Comment (jiazhihao) [inline python/mirage/mpk/models/qwen3/modeling_qwen3.py:1]: It seems Qwen3Model was constructed but not used much to construct megakernels. I was wondering if we can remove `modeling_qwen3.py` and only use `configuration_qwen3.py` for constructing megakernel. This change would make our model layer very lightweight---only a configuration file and a builder file.

File: demo/qwen3/demo.py [hunk 3]
```diff
@@ -254,7 +254,7 @@ def max_factor_leq_n(m: int, n: int) -> int:
         else:
             profiler_tensor = None
             
-        spec_decode_config = mi.speculative.spec_decode_class(
+        spec_decode_config = mi.mpk.spec_decode_class(
```
Comment (JackFram) [inline demo/qwen3/demo.py:257]: Did we test with speculative decoding serving cases?

File: demo/qwen3/demo_30B_A3B.py
Comment (JackFram) [inline demo/qwen3/demo_30B_A3B.py:1]: Might also need to change the spec_config init to the updated version

File: demo/qwen3/demo_mpk_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    
+
+    if args.profiling:
+        profiler_tensor = torch.zeros(
+            3000 * 128, dtype=torch.uint64, device="cuda"
+        ).contiguous()
+    else:
+        profiler_tensor = None
+        
+    spec_decode_config = mi.mpk.speculative.spec_decode_class(
```
Comment (JackFram) [inline demo/qwen3/demo_mpk_wrapper.py:101]: can do `spec_decode_config = mi.mpk.spec_decode_class(` directly?

**Review by JackFram — CHANGES_REQUESTED** [review]
I have tested on B200 that the demo works, should be in good shape, left some comments on small changes to be addressed before merging.
